doc/cephfs: recover file system after catastrophic monitor store loss (#42295)
ceph/ceph: doc/cephfs/disaster-recovery-experts.rst

    will contain holes in place of the missing data.

.. _Symbolic link recovery: https://tracker.ceph.com/issues/46166

Restoring file system after recovering monitor stores using OSDs
----------------------------------------------------------------

During rare occasions, all the monitor stores of a cluster may get corrupted.
To recover the cluster in such a scenario, you need to rebuild the monitor
stores using the OSDs, and get back the pools intact (active+clean state).

> However, the rebuilt monitor stores don't restore the FS/MDS maps. Additonal

Reviewer: ```suggestion
However, the rebuilt monitor stores don't restore the FS/MDS maps. Additional
```

steps are required to bring back the file system. Currently, only the steps to
recover a **single active MDS** file system have been identified, and tested.

> These are described below.

Reviewer: add note about multiple file system

> 

Reviewer: I think it would be helpful to outline the general idea of the recovery: we're creating the FSMap with some basic defaults (we cannot necessarily recovery every setting applied) and allowing the MDS to recover from what is in the journal/metadata of the pools.

> Stop all the MDSs of the cluster.

Reviewer: Provide steps to verify no MDS are running: `ceph fs dump`.

.. warning::

     If the MDSs are not stopped, they may start wiping the pools when

>      recreating the file system.

Reviewer: ??

Author: Yeah, wiping pools was a lazy way of saying it. So if the MDSs are still up, while the `fs new` command is executed, I think the MDSs don't pick up the existing in-RADOS metadata and start overwriting the metadata. As mentioned here, https://github.com/ceph/ceph/blob/master/doc/cephfs/disaster-recovery-experts.rst#mds-map-reset
"...  is that doing a remove/new will leave rank 0 in 'creating' state, such that it would overwrite any existing root inode on disk and orphan any existing files. In contrast, the 'reset' command will leave rank 0 in 'active' state such that the next MDS daemon to claim the rank will go ahead and use the existing in-RADOS metadata."""

Recreate the file system using the recovered file system pools.

::

    ceph fs new <fs_name> <metadata_pool> <data_pool> --force

The file system cluster ID, fscid, of the file system will not be preserved.
To preserve the fscid, pass the desired fscid when recreating the file system.

::

    ceph fs new <fs_name> <metadata_pool> <data_pool> --fscid <fscid> --force

Reset the file system.

::

>     ceph fs reset <fs_name> --yes-i-really-mean-it

Reviewer: Do we know specifically what this is for? Is it necessary? I think resetting a new file system is redundant.

Author: It's necessary. I tried restarting the MDSs without executing the steps. While the MDSs do come up, and filesystem status is OK. the data in the file system wasn't retrieved. I think it might be tbecauserank 0 is not set to failed state (which is done by `fs reset` command  https://github.com/ceph/ceph/blob/master/src/mds/FSMap.cc#L503 ) and is in creating state, so when an MDS picks up rank 0 it starts overwriting the metadata instead of fetching the existing metadata?

Reviewer: Ah. This seems like an unsatisfactory situation to me. We should provide an option for `fs new` to start rank 0 in a failed state so that this `fs reset` is not necessary. Requiring all MDS to be turned off is onerous in a multifs system.

Let's keep the doc as-is but please add a tracker ticket to fix that.

Author: Got it. Let me do this.

Author: @batrick created https://tracker.ceph.com/issues/51716
Please check if it looks OK to you.

Restart the MDSs. Check that the file system is no longer in degraded state and

> one of the MDSs is active.

Reviewer: Add some notes about what's not recovered. At the very least:

- additional data pools must be added manually
- standby_count_wanted
- client feature requirements

    will contain holes in place of the missing data.

.. _Symbolic link recovery: https://tracker.ceph.com/issues/46166

> Restoring a single active MDS file system after recovering monitor stores using OSDs

Reviewer: ```suggestion
Recovering the file system after catastrophic Monitor store loss
```

====================================================================================

> During rare occasions, all the monitor stores of a cluster may get corrupted.

Reviewer: ```suggestion
During rare occasions, all the monitor stores of a cluster may get corrupted or lost.
```

To recover the cluster in such a scenario, you need to rebuild the monitor

> stores using the OSDs, and get back the pools intact (active+clean state).

Reviewer: Can we link to an existing RADOS doc about this?

> However, the rebuilt monitor stores don't restore the FS/MDS maps. Additional

Reviewer: ```suggestion
However, the rebuilt monitor stores don't restore the file system maps ("FSMap"). Additional
```

steps are required to bring back the file system. The steps to recover a

> multiple active MDS file system, and multiple file systems are yet to be

Reviewer: ```suggestion
multiple active MDS file system or multiple file systems are yet to be
```

identified. Currently, only the steps to recover a **single active MDS** file

> system with no additional file systems in the cluster have been identified,

Reviewer: ```suggestion
system with no additional file systems in the cluster have been identified
```

and tested. Briefly the steps are: stop the MDSs; recreate the FSMap with
basic defaults; and allow MDSs to recover from the journal/metadata stored
in the filesystem's pools. The steps are described in more detail below.

First up, stop all the MDSs of the cluster.

Verify that the MDSs have been stopped. Execute the below command and
check that no active or standby MDS daemons are listed for the file system.

::

    ceph fs dump

Recreate the file system using the recovered file system pools. The new FSMap
will have the filesystem's default settings. However, the user defined file
system settings such as ``standby_count_wanted``, ``required_client_features``,
extra data pools, etc., are lost and need to be reapplied later.

::

    ceph fs new <fs_name> <metadata_pool> <data_pool> --force

The file system cluster ID, fscid, of the file system will not be preserved
automatically. Pass the desired fscid when recreating the file system to do

> so.

Reviewer: Explain when creating a file system with the old fscid is desirable.

::

    ceph fs new <fs_name> <metadata_pool> <data_pool> --fscid <fscid> --force

Next, reset the file system. The below command marks the state of the

> filesystem's rank 0 such that eventually when a MDS daemon picks up rank 0 the

Reviewer: ```suggestion
file system's rank 0 such that eventually when a MDS daemon picks up rank 0 the
```

daemon reads the existing in-RADOS metadata and doesn't overwrite it.

::

    ceph fs reset <fs_name> --yes-i-really-mean-it

Restart the MDSs. Check that the file system is no longer in degraded state and
one of the MDSs is active.

::

    ceph fs dump

> Reapply the custom file system settings.

Reviewer: ```suggestion
Reapply any other custom file system settings.
```
